ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh

```diff
@@ -0,0 +1,167 @@
+
+  /**
+   * \brief Computes the stresses in the Muesli finite strain material model.
+   * \tparam voigt A boolean indicating whether to return stresses in Voigt notation.
+   * \tparam Derived The derived type of the input matrix.
+   * \param C The right Cauchy-Green tensor.
+   * \return StressMatrix The stresses.
+   */
+  template <bool voigt, typename Derived>
+  auto stressesImpl(const Eigen::MatrixBase<Derived>& C) const {
+    static_assert(Concepts::EigenMatrixOrVoigtNotation3<Derived>);
+    if constexpr (!voigt) {
+      if constexpr (!Concepts::EigenVector<Derived>) {
+        updateState(C);
+        mp_->secondPiolaKirchhoffStress(stress_);
+        return Muesli::toMatrix(stress_);
+      } else
+        static_assert(!Concepts::EigenVector<Derived>,
+                      "MuesliFiniteStrain can only be called with a matrix and not a vector in Voigt notation");
+    } else
+      static_assert(voigt == false, "MuesliFiniteStrain does not support returning stresses in Voigt notation");
+  }
+
+  /**
+   * \brief Computes the tangent moduli in the Muesli finite strain material model.
+   * \tparam voigt A boolean indicating whether to return tangent moduli in Voigt notation.
+   * \tparam Derived The derived type of the input matrix.
+   * \param C The right Cauchy-Green tensor.
+   * \return Eigen::TensorFixedSize<ScalarType, Eigen::Sizes<3, 3, 3, 3>> The tangent moduli.
+   */
+  template <bool voigt, typename Derived>
+  auto tangentModuliImpl(const Eigen::MatrixBase<Derived>& C) const {
+    static_assert(Concepts::EigenMatrixOrVoigtNotation3<Derived>);
+    if constexpr (!voigt) {
+      if constexpr (!Concepts::EigenVector<Derived>) {
+        updateState(C);
+
+        mp_->convectedTangent(tangentModuli_);
+        return Muesli::toTensor(tangentModuli_);
+      } else
+        static_assert(!Concepts::EigenVector<Derived>,
+                      "MuesliFiniteStrain can only be called with a matrix and not a vector in Voigt notation");
+    } else
+      static_assert(voigt == false, "MuesliFiniteStrain does not support returning tangent moduli in Voigt notation");
+  }
+
+  /**
+   * \brief Returns the underlying muesli material implementation.
+   * \return auto& reference to the muesli material.
+   */
+  auto& material() const { return material_; }
+
+  /**
+   * \brief asserts that the materialpoint pointer is not null.
+   */
+  bool assertMP() const { return mp_.get() != NULL; }
+
+  FiniteStrain(const FiniteStrain& other)
+      : materialParameter_{other.materialParameter_},
+        material_{Dune::className<FM>(), materialParameter_},
+        mp_{material_.createMaterialPoint()} {}
+
+private:
+  MaterialParameters materialParameter_;
+  MaterialModel material_;
+  std::unique_ptr<muesli::finiteStrainMP> mp_;
+
+  mutable istensor strain_{};
+  mutable istensor stress_{};
+  mutable itensor4 tangentModuli_{};
+
+  template <typename Derived>
+  void updateState(const Eigen::MatrixBase<Derived>& C) const {
```

Comment:
A method called updateState should not be marked `const`. If the strain objects are directly Eigen types then you also do not need the evil mutable variables.

Reply:
I can of course get rid of the mutable variables, my motivation was to avoid allocating at each evaluation